torchdp/per_sample_gradient_clip.py

```diff
@@ -125,39 +126,55 @@ def calc_thresh_value(self, layer_norms):
             thresh_norm if self.clip_per_layer else thresh_norm * len(self.named_params)
         )
 
-    def clip(self) -> List[float]:
+    def clip(self, virtual: bool = False) -> (List[float], int):
         r"""
         Clips the grad_sample stored in .grad_sample by computing a per-sample
         norm clip factor, using it to rescale each sample's gradient in
         .grad_sample to norm clip, then averaging them back into .grad.
+        If 'virtual' is set, the clipped gradients are summed up into a
+        temporary accumulator instead.
 
         The gradients of the model's parameters are modified in-place.
 
         We assume the batch size is the first dimension.
 
         Arguments:
-            tensor (Tensor): a single Tensor whose norm will be normalized
+            virtual (bool): if set, the clipped gradients in this mini-batch
+            are summed up into a accummulator for a larger batch
 
         Returns:
-            A dictionary of parameter names and their clipped norm
+            A list of clipping thresholds
         """
+
+        # check if we've already accumulated all the clipped gradients for this batch
+        if len(self.accumulation_state) > 0 and not virtual:
+            return self.finalize_batch()
+
         # step 0 : calculate the layer norms and thresholds
         all_norms = self.get_all_layer_norms()
         thresh_norms = self.calc_thresh_value(all_norms)
+        batch_size = len(thresh_norms[0][1])
         threshs = []
         for thresh_norm, named_param in zip(thresh_norms, self.named_params):
             # step 1 : Find the clipping factor per layer (per parameter set)
             thresh, norm = thresh_norm
             per_sample_clip_factor = thresh / (norm + 1e-6)
             # We are *clipping* the gradient, so if the factor is ever >1 we set it to 1
             per_sample_clip_factor = per_sample_clip_factor.clamp(max=1.0)
-            b_sz = len(per_sample_clip_factor)  # all batch sizes are the same
             # step 2: Do the clipping
             name, p = named_param
             pre_clip_pos = p.grad_sample.mean(0) > 0
-            p.grad = (
-                torch.einsum("i,i...", per_sample_clip_factor, p.grad_sample) / b_sz
-            )
+            summed_grad = torch.einsum("i,i...", per_sample_clip_factor, p.grad_sample)
```

Comment:
Do we need p.grad_sample after this line anymore?
Could we remove the clear_grad_sample function and do this here?

Reply:
many tests in layers_grad_test.py assume that p.grad_sample will still exist after we've called clipper.step(). So at the very least this would require changing all those tests.

Also, we'd still need clear_grad_sample() because it has to be called in optimizer.zero_grad(): if the effective batch size doesn't evenly divide the training set size, the last effective batch in a training epoch is incomplete: we accumulate a bunch of minibatches, but never call .step()
By calling .zero_grad() at the start of an epoch, we ensure that all the accumulated gradients are deleted.

Reply:
also some comments below on some other cleanings.